algorithms/src/snark/marlin/ahp/ahp.rs

```diff
@@ -319,6 +376,22 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
         lhs *= selector_at_gamma;
         lhs
     }
+
+    fn get_selector_evaluation(
+        cached_selector_evaluations: &mut BTreeMap<u64, F>,
+        largest_non_zero_domain: &EvaluationDomain<F>,
+        target_domain: &EvaluationDomain<F>,
+        gamma: F,
+    ) -> F {
+        match cached_selector_evaluations.get(&target_domain.size) {
+            Some(selector) => *selector,
+            None => {
+                let selector = largest_non_zero_domain.evaluate_selector_polynomial(*target_domain, gamma);
+                cached_selector_evaluations.insert(target_domain.size, selector);
+                selector
+            },  
+        }
+    }
```

Comment:
I think we can use the `entry` API here: https://doc.rust-lang.org/std/collections/btree_map/enum.Entry.html

Reply:
Very fancy!